File: rust/src/aoc2022/day4.rs
use rdcl_aoc_helpers::input::WithReadLines;
use std::fs::File;

pub fn day4(path: &String) {
    println!("Day 4");
    let data:Vec<String> = File::open(path).read_lines::<String>(1).collect();
    let mut nb_elves_over:u32 = 0;
    let mut nb_elves_inter = 0;
    for row in data{
        let pair:Vec<u32> = row.split(&['-', ','][..]).map(|s| s.parse::<u32>().unwrap()).collect();
        let res = overwork(pair[0], pair[1], pair[2], pair[3]);
        nb_elves_over += res[0];
        nb_elves_inter += res[1];
    }
    
    println!("Part 1: {}", nb_elves_over);
    println!("Part 2: {}", nb_elves_inter);
}

fn overwork(n1:u32, n2: u32, m1: u32, m2: u32) -> Vec<u32> {
    let mut full = 0;
    let mut partial = 0;
    if ((n1 <= m1 && m1 <= n2) && (n1 <= m2 && m2 <= n2)) || ((m1 <= n1 && n1 <= m2) && (m1 <= n2 && n2 <= m2)){
        full = 1;
    }
    if ((n1 <= m1 && m1 <= n2) || (n1 <= m2 && m2 <= n2)) || ((m1 <= n1 && n1 <= m2) || (m1 <= n2 && n2 <= m2)){
        partial = 1;
    }
    return vec![full, partial]
}

File: python/aoc2022/days/day4.py
def day4():
    with open('../input/input4.txt', 'r') as f:
        data = f.read().split('\n')[:-1]
        pairs = [(n.split(',')[0], n.split(',')[1]) for n in data]

    nb_idle_elves = 0
    nb_over_elves = 0

    for pair in pairs:
        f, p = overlap(pair[0], pair[1])
        nb_idle_elves += f
        nb_over_elves += p

    print('Part 1: ', nb_idle_elves)
    print('Part 2: ', nb_over_elves)


def overlap(pair1, pair2):
    full = partial = 0
    n = pair1.split('-')
    n1, n2 = int(n[0]), int(n[1])
    m = pair2.split('-')
    m1, m2 = int(m[0]), int(m[1])
    if ((n1 <= m1 <= n2) and (n1 <= m2 <= n2)) or ((m1 <= n1 <= m2) and (m1 <= n2 <= m2)):
        full = 1
    if ((n1 <= m1 <= n2) or (n1 <= m2 <= n2)) or ((m1 <= n1 <= m2) or (m1 <= n2 <= m2)):
        partial = 1
    return full, partial
